featurebyte/api/user_defined_function_injector.py

```diff
@@ -0,0 +1,331 @@
+            node_type=NodeType.GENERIC_FUNCTION,
+            node_params=node_params.dict(by_alias=True),
+            node_output_type=NodeOutputType.SERIES,
+            input_nodes=input_nodes,
+        )
+        return node
+
+    @staticmethod
+    def _generate_docstring(udf: UserDefinedFunctionModel) -> str:
+        parameters = "\n        ".join(
+            [
+                f"{param.name} : {get_param_type_annotations(param.dtype)}"
+                for param in udf.function_parameters
+            ]
+        )
+        docstring_template = f"""
+        This function uses the feature store SQL function `{udf.function_name}` to generate a new column
+        or new feature. Note that the non-scalar input parameters must be homogeneous, i.e., they must
+        be either all view columns or all features. The output type is determined by the input type.
+        If all input parameters are view columns, then the output type is a view column. If all input
+        parameters are features, then the output type is a feature. All view columns input parameters
+        must have matching row alignment. Otherwise, an error is raised.
+
+        Parameters
+        ----------
+        {parameters}
+
+        Returns
+        -------
+        Union[ViewColumn, Feature]
+            A new column or new feature generated by the feature store function `{udf.function_name}`.
+        """
+        return textwrap.dedent(docstring_template).strip()
+
+    @classmethod
+    def create_and_add_function(
+        cls, func_accessor: FunctionAccessor, udf: UserDefinedFunctionModel
+    ) -> None:
+        """
+        Create a dynamic method based on the user defined function and add it to the function accessor.
+
+        Parameters
+        ----------
+        func_accessor: FunctionAccessor
+            The function accessor instance which contains the dynamic method
+        udf: UserDefinedFunctionModel
+            The user defined function model which contains the function information
+        """
+
+        def _method_wrapper(*args: Any, **kwargs: Any) -> Union[ViewColumn, Feature]:
+            param_processor = FunctionParameterProcessor(udf.function_parameters)
+            extracted_params = param_processor.process(*args, **kwargs)
+            node = cls._insert_generic_function_node(
+                function_parameters=extracted_params.node_function_parameters,
+                input_node_names=extracted_params.input_node_names,
+                udf=udf,
+            )
+            return extracted_params.output_type(
+                name=None,
+                node_name=node.name,
+                feature_store=extracted_params.feature_store,
+                tabular_source=extracted_params.tabular_source,
+                dtype=udf.output_dtype,
+            )
+
+        # assign the dynamic method to the function accessor
+        assert udf.name is not None
+        setattr(func_accessor, udf.name, _method_wrapper)
```

Comment:
Is there already some validation that the udf name is a valid python identifier?

Reply:
Good catch. Current validation is only done the `function_name` but not `name`.